### ISSUE TITLE
Does Identity API support Keystone V3?

### ISSUE THREAD

**Issue by westenb** [ContainX/openstack4j#645]
Hello,

does the Identity API support Keystone V3 for e.g. accessing users / projects / etc. In general, the identity() API of openstack4j seems to be Keystone V2 based in terms of terminology.

In my case, I am trying to access the identity service as follows (using OpenStack4j 2.11):
os.identity().users().getByName(aclUser)

and I am getting in my OpenStack environment with Keystone V3 the following exception pointing to some format incompatibility:

```
com.fasterxml.jackson.databind.JsonMappingException: Root name 'users' does not match expected ('user') for type [simple type, class org.openstack4j.openstack.identity.domain.KeystoneUser]
 at [Source: org.glassfish.jersey.message.internal.EntityInputStream@75fc8df; line: 1, column: 2]
        at com.fasterxml.jackson.databind.JsonMappingException.from(JsonMappingException.java:148)
        at com.fasterxml.jackson.databind.ObjectReader._unwrapAndDeserialize(ObjectReader.java:1439)
        at com.fasterxml.jackson.databind.ObjectReader._bind(ObjectReader.java:1229)
        at com.fasterxml.jackson.databind.ObjectReader.readValue(ObjectReader.java:676)
        at com.fasterxml.jackson.jaxrs.base.ProviderBase.readFrom(ProviderBase.java:777)
        at org.glassfish.jersey.message.internal.ReaderInterceptorExecutor$TerminalReaderInterceptor.invokeReadFrom(ReaderInterceptorExecutor.java:257)
        at org.glassfish.jersey.message.internal.ReaderInterceptorExecutor$TerminalReaderInterceptor.aroundReadFrom(ReaderInterceptorExecutor.java:229)
        at org.glassfish.jersey.message.internal.ReaderInterceptorExecutor.proceed(ReaderInterceptorExecutor.java:149)
        at org.glassfish.jersey.message.internal.MessageBodyFactory.readFrom(MessageBodyFactory.java:1124)
        at org.glassfish.jersey.message.internal.InboundMessageContext.readEntity(InboundMessageContext.java:851)
        at org.glassfish.jersey.message.internal.InboundMessageContext.readEntity(InboundMessageContext.java:783)
        at org.glassfish.jersey.client.ClientResponse.readEntity(ClientResponse.java:326)
        at org.glassfish.jersey.client.InboundJaxrsResponse$1.call(InboundJaxrsResponse.java:111)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:315)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:297)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:228)
        at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:399)
        at org.glassfish.jersey.client.InboundJaxrsResponse.readEntity(InboundJaxrsResponse.java:108)
        at org.openstack4j.connectors.jersey2.HttpResponseImpl.readEntity(HttpResponseImpl.java:113)
        at org.openstack4j.core.transport.HttpEntityHandler.handle(HttpEntityHandler.java:65)
        at org.openstack4j.core.transport.HttpEntityHandler.handle(HttpEntityHandler.java:24)
        at org.openstack4j.connectors.jersey2.HttpResponseImpl.getEntity(HttpResponseImpl.java:63)
        at org.openstack4j.openstack.internal.BaseOpenStackService$Invocation.execute(BaseOpenStackService.java:189)
        at org.openstack4j.openstack.internal.BaseOpenStackService$Invocation.execute(BaseOpenStackService.java:183)
        at org.openstack4j.openstack.identity.internal.UserServiceImpl.getByName(UserServiceImpl.java:104)
```

Regards, Eric

**Comment by vinodborole** [comment]
@westenb It seems the osClient you got is not of v3 as the keystoneuser class selected is of v2. Following is the code how you get the v3 token

               os = OSFactory.builderV3().useNonStrictSSLClient(false).credentials("user", "passwd").scopeToProject(Identifier.byId("352532532532532"),Identifier.byName("default")).endpoint("http://openstack:5000/v3").authenticate();

**Comment by westenb** [comment]
Hi,

actually my osClient is of type V3:
```
OSFactory.builderV3().endpoint(config.getAuthUrl() + "/v3")  .scopeToProject(Identifier.byName(config.getTenantName()),Identifier.byName(config.getDomainName()))
  .credentials(config.getUserName(), config.getPassword())
  .withConfig(transportConfig)
  .authenticate();
```
and it works with our OpenStack setup which only supports Keystone V3, and I can access e.g. SWIFT. However, I don't see how the V3 identity service is supported (if indeed it is).
The API method identity() returns a instance of IdentityService (independently of whether it is Keystone V2 or V3). 

For example via this instance I get access to the TenantService which retrieves the tenant list via /tenants, and not /projects as documented here http://developer.openstack.org/api-ref-identity-v3.html#projects-v3

It seems to me that there might be a need to provide a Keystone V3 specific service API, e.g. as in the following pseudo code
```
IdentityServiceV3 service = os.identityV3();
```
but probably this would cascade down to all contained interfaces (users, projects, endpoints). However, due to the massive incompabilities / changes between Keystone V2 and V3; I am not sure how to hide this in one API.

Regards,
Eric

**Comment by vinodborole** [comment]
@westenb I think it still in development and should be available in the next release, you refer to this branch for v3 latest development

https://github.com/gondor/openstack4j/tree/v3_collaboration

**Comment by westenb** [comment]
@vinodborole thanks; will have a look.

Maybe another more general question: I am struggeling a bit with the versioning of openstack4j. Latest released version in Maven central is 2.11, but actually in GitHub there is no corresponding tag / release. The master branch is at 2.20-SNAPSHOT, and the V3 is - I assume - the early development for a new major version. Is there somewhere a kind of release roadmap available?

**Comment by vinodborole** [comment]
@westenb I am sorry i do not have that information. May be @gondor can answer that.

**Comment by auhlig** [comment]
@westenb: It's not in early development :) 
I think you can expect a new release of Os4j supporting Identity V3 soon.

**Comment by gondor** [comment]
@westenb We hopefully be releasing V3 by the end of this week or latest early next week.

**Comment by auhlig** [comment]
@westenb: I guess the issue can be closed as Os4j 3.0.0 supports V3 .
